Can no longer build GHC with Musl

GHC can no longer be built with the Musl libc because the `ghc822Binary` compiler needed to bootstrap no longer passes the install test. This prevents fully static executables from being produced with this tool chain.

I've `git bisect`ed and 0450d28a (cc @dtzWill) was the commit that introduced the regression. To reproduce:

```sh
$ NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs/archive/0450d28a.tar.gz nix-build '<nixpkgs>' -A pkgsMusl.haskell.compiler.ghc822Binary --no-out-link
(....)
running install tests
[1 of 1] Compiling Main             ( main.hs, main.o )
Linking main ...
/nix/store/wvj2kzhypkkwqx6qjcghm60bz6pqlqzz-binutils-2.31.1/bin/ld: /nix/store/f0061cv30fk6d2lxrivakbwqk0fqvlgp-ghc-8.2.2-binary/lib/ghc-8.2.2/rts/libHSrts.a(Hpc.o): in function `startupHpc':

/home/ben/bin-dist-8.2.2-Linux/ghc/rts/Hpc.c:214:0: error:
     undefined reference to `__strdup'
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)
builder for '/nix/store/i3s9a1zpma6pscf6xdi91r4h0hvyy87a-ghc-8.2.2-binary.drv' failed with exit code 1
error: build of '/nix/store/i3s9a1zpma6pscf6xdi91r4h0hvyy87a-ghc-8.2.2-binary.drv' failed
```
The previous commit, b262be32, works:

```sh
NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs/archive/b262be32.tar.gz nix-build '<nixpkgs>' -A pkgsMusl.haskell.compiler.ghc822Binary --no-out-link
(...)
running install tests
[1 of 1] Compiling Main             ( main.hs, main.o )
Linking main ...
compilation ok
/nix/store/pxd37xckl30yfky2rbj3a7bhz9fgvjkl-ghc-8.2.2-binary
```

--- a/pkgs/development/compilers/ghc/8.2.2-binary.nix
+++ b/pkgs/development/compilers/ghc/8.2.2-binary.nix
@@ -115,7 +115,7 @@ stdenv.mkDerivation rec {
     # [1] http://refspecs.linuxbase.org/LSB_4.0.0/LSB-Core-generic/LSB-Core-generic/baselib---strdup-1.html
     # Use objcopy magic to make the change:
     stdenv.lib.optionalString stdenv.hostPlatform.isMusl ''
-      find ./ghc-${version}/rts -name "libHSrts*.a" -exec objcopy --redefine-sym __strdup=strdup {} \;
+      find ./ghc-${version}/rts -name "libHSrts*.a" -exec ''${OBJCOPY:-objcopy} --redefine-sym __strdup=strdup {} \;
     '';
 
   configurePlatforms = [ ];

--- a/pkgs/development/compilers/ghc/8.2.2-binary.nix
+++ b/pkgs/development/compilers/ghc/8.2.2-binary.nix
@@ -107,6 +107,15 @@ stdenv.mkDerivation rec {
 
       sed -i "s|/usr/bin/perl|perl\x00        |" ghc-${version}/ghc/stage2/build/tmp/ghc-stage2
       sed -i "s|/usr/bin/gcc|gcc\x00        |" ghc-${version}/ghc/stage2/build/tmp/ghc-stage2
+    '' +
+    # We're kludging a glibc bindist into working with non-glibc...
+    # Here we patch up the use of `__strdup` (part of glibc binary ABI)
+    # to instead use `strdup` since musl doesn't provide __strdup
+    # (`__strdup` is defined to be an alias of `strdup` anyway[1]).
+    # [1] http://refspecs.linuxbase.org/LSB_4.0.0/LSB-Core-generic/LSB-Core-generic/baselib---strdup-1.html
+    # Use objcopy magic to make the change:
+    stdenv.lib.optionalString stdenv.hostPlatform.isMusl ''
+      find ./ghc-${version}/rts -name "libHSrts*.a" -exec objcopy --redefine-sym __strdup=strdup {} \;
     '';
 
   configurePlatforms = [ ];

--- a/pkgs/development/compilers/ghc/8.6.3-binary.nix
+++ b/pkgs/development/compilers/ghc/8.6.3-binary.nix
@@ -99,6 +99,15 @@ stdenv.mkDerivation rec {
 
       sed -i "s|/usr/bin/perl|perl\x00        |" ghc-${version}/ghc/stage2/build/tmp/ghc-stage2
       sed -i "s|/usr/bin/gcc|gcc\x00        |" ghc-${version}/ghc/stage2/build/tmp/ghc-stage2
+    '' +
+    # We're kludging a glibc bindist into working with non-glibc...
+    # Here we patch up the use of `__strdup` (part of glibc binary ABI)
+    # to instead use `strdup` since musl doesn't provide __strdup
+    # (`__strdup` is defined to be an alias of `strdup` anyway[1]).
+    # [1] http://refspecs.linuxbase.org/LSB_4.0.0/LSB-Core-generic/LSB-Core-generic/baselib---strdup-1.html
+    # Use objcopy magic to make the change:
+    stdenv.lib.optionalString stdenv.hostPlatform.isMusl ''
+      find ./ghc-${version}/rts -name "libHSrts*.a" -exec objcopy --redefine-sym __strdup=strdup {} \;
     '';
 
   configurePlatforms = [ ];

--- a/pkgs/development/compilers/ghc/8.6.3-binary.nix
+++ b/pkgs/development/compilers/ghc/8.6.3-binary.nix
@@ -107,7 +107,7 @@ stdenv.mkDerivation rec {
     # [1] http://refspecs.linuxbase.org/LSB_4.0.0/LSB-Core-generic/LSB-Core-generic/baselib---strdup-1.html
     # Use objcopy magic to make the change:
     stdenv.lib.optionalString stdenv.hostPlatform.isMusl ''
-      find ./ghc-${version}/rts -name "libHSrts*.a" -exec objcopy --redefine-sym __strdup=strdup {} \;
+      find ./ghc-${version}/rts -name "libHSrts*.a" -exec ''${OBJCOPY:-objcopy} --redefine-sym __strdup=strdup {} \;
     '';
 
   configurePlatforms = [ ];

